### ISSUE TITLE
Support PHP8.1

### ISSUE THREAD

**Issue by ChristophWurst** [nextcloud/server#29287]
According to https://php.watch/news/2021/03/php81-release-date PHP8.1 will be released by end of next month. That will mean the release happens somewhere around the release of Nextcloud 23 and we can assume that the early adopters will soon update their systems and find out Nextcloud 23 doesn't support 8.1. Ref https://github.com/nextcloud/server/blob/822623109f4f4b3252c04a5c0ccf6757a882e771/lib/versioncheck.php#L36-L42.

As far as I know php8.1 is "just" a minor release but some libs/code have to be adjusted. We'll have to decide whether to look into this for 23, 23.1 or 24.

Potential breaking changes
* Mysqli error handling https://kinsta.com/blog/php-8-1/#mysqli-default-error-mode-set-to-exceptions – possibly taken care of in our db abstraction libs but we'll have to make sure we install a compatible version

cc @AndyScherzinger @juliushaertl @nickvergessen @skjnldsv

**Comment by ChristophWurst** [comment]
I checked the 3rdparty repo and no dep is actively blocking php8.1 right now (overrode the version number to test this)

**Comment by AndyScherzinger** [comment]
thanks for the ping @ChristophWurst given the amount of time and people I'd not rush this for 23.0.0 also given the fact that apps might just break because they have php8.1 issues and that covers ours as well as community apps. So this seems to be to short notice for app devs and likely is also to risky for us and the v23 timeline given that we already reach beta status.

**Comment by ChristophWurst** [comment]
> Does anyone know the reason for [d690f90](https://github.com/nextcloud/server/commit/d690f909284ae4bb4dee7d00318104ee76720bfa) ?

https://github.com/nextcloud/server/pull/23780#issuecomment-739883843

I suggest we revert ans see what happens :)

**Comment by come-nc** [comment]
The tests for Archive run into this problem: https://stackoverflow.com/questions/64698935/using-ziparchive-with-php-8-and-temporary-files
I tried fixing it by this change:
```
diff --git a/tests/lib/Archive/ZIPTest.php b/tests/lib/Archive/ZIPTest.php
index 17a639c9f5..b6eb947f53 100644
--- a/tests/lib/Archive/ZIPTest.php
+++ b/tests/lib/Archive/ZIPTest.php
@@ -17,6 +17,8 @@ class ZIPTest extends TestBase {
        }
 
        protected function getNew() {
-               return new ZIP(\OC::$server->getTempManager()->getTemporaryFile('.zip'));
+               $tmpfile = \OC::$server->getTempManager()->getTemporaryFile('.zip');
+               \unlink($tmpfile);
+               return new ZIP($tmpfile);
        }
 }
```

But I get a misterious warning at the end of the tests:
```
PHP Warning:  PHPUnit\TextUI\Command::main(): Cannot destroy the zip context: Can't remove file: No such file or directory in /home/mcmic/.config/composer/vendor/phpunit/phpunit/src/TextUI/Command.php on line 96
```

No idea if the actual code is concerned.

**Comment by come-nc** [comment]
> `libxml_disable_entity_loader` is deprecated but I do not understand what it does and why we call it. Leaving it in for now.

https://php.watch/versions/8.0/libxml_disable_entity_loader-deprecation

**Comment by come-nc** [comment]
- [ ] giggsey/libphonenumber-for-php should be bumped to at least 8.12.34 https://github.com/giggsey/libphonenumber-for-php/releases
- [ ] scssphp/scssphp should be bumped to 1.8.1 https://github.com/scssphp/scssphp/releases
- [ ] swiftmailer should be bumped to 6.3.0, but it’s abandoned at the end of the month, so moving to symfony/mailer should be looked into https://github.com/swiftmailer/swiftmailer

**Comment by come-nc** [comment]
Opened https://github.com/nextcloud/server/issues/29731 about swiftmailer situation.

**Comment by come-nc** [comment]
8.1 is officially out.
Highlights: https://www.php.net/releases/8.1/en.php
Changelog: https://www.php.net/ChangeLog-8.php#8.1.0
Migration guide: https://www.php.net/manual/en/migration81.php

**Comment by come-nc** [comment]
Regarding
```
FILE: /home/mcmic/dev/nextcloud/server/lib/private/L10N/Factory.php
-----------------------------------------------------------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
-----------------------------------------------------------------------------------------------------------------------------------
 595 | ERROR | Function create_function() is deprecated since PHP 7.2 and removed since PHP 8.0; Use an anonymous function instead
 600 | ERROR | Function create_function() is deprecated since PHP 7.2 and removed since PHP 8.0; Use an anonymous function instead
-----------------------------------------------------------------------------------------------------------------------------------
```

It appears that both calls are in `OC\L10N\Factory::createPluralFunction` which is never called in core, and is not tested by FactoryTest.php.

Is the method safe to remove or may it be used by applications?

**Comment by ChristophWurst** [comment]
Seems unused https://github.com/search?q=org%3Anextcloud+createPluralFunction&type=code

Nevertheless this is a public API so we can't "just" remove it.

**Comment by nickvergessen** [comment]
Should have never been public in first place from my pov.
Especially since https://github.com/nextcloud/server/pull/26375
I'm fine with removing directly or making it throw an exception instead

**Comment by come-nc** [comment]
> Should have never been public in first place from my pov. Especially since #26375 I'm fine with removing directly or making it throw an exception instead

Can you create the PR for that then?

**Comment by nickvergessen** [comment]
No, I'm on vacation until next year

**Comment by txt-file** [comment]
As debian started the php7.4 → php8.1 transition on 2021-12-31 I stumbled over nextclouds missing php8.1 support.
I would be happy to drop the transitional php7.4 and move to 8.1 soon. :-)

**Comment by nickvergessen** [comment]
8.1 support will only vome with Nextcloud 24 anyway

**Comment by bcutter** [comment]
Even NC 22 with PHP 8.0 after updating to 8.1 (but not removing 8.0) already creates issues - like the cron job not working anymore.

OCC commands give

`This version of Nextcloud is not compatible with > PHP 8.0.<br/>You are currently running 8.1.1.`

So it seems like parts of NC already (try to) use PHP 8.1.

What a mess -.-
Think I will need to downgrade PHP because NC is not fast enough.

**Has NC 23 PHP 8.1 support?**

**Comment by come-nc** [comment]
> Even NC 22 with PHP 8.0 after updating to 8.1 (but not removing 8.0) already creates issues - like the cron job not working anymore.
> 
> Also OCC commands give
> 
> `This version of Nextcloud is not compatible with > PHP 8.0.<br/>You are currently running 8.1.1.`
> 
> So it seems like parts of NC already (try to) use PHP 8.1.
> 
> What a mess -.- Think I will need to downgrade PHP because NC is not fast enough.
> 
> **Has NC 23 PHP 8.1 support?**
> 
> Update: no obviously not according to https://docs.nextcloud.com/server/latest/admin_manual/installation/system_requirements.html so upgrading NC to v23 is not an option. Need to stay on PHP 8.0... argh confounded

PHP 8.1 is not supported by Nextcloud yet (which is why this issue is "open").
It should be supported by 24 when it’s out, maybe also by some future 22/23 minor versions, not sure yet.

If you have both 8.0 and 8.1 on your system, it’s your job to make sure you run Nextcloud and occ (and the cron job) with 8.0. Otherwise you will get the compatibility error as expected.

**Comment by bcutter** [comment]
Figured that out meanwhile too. Used "update-alternatives" to redirect NC to the 8.0 PHP instance for the time being.

**Comment by pooh22** [comment]
the update-alternatives trick didn't work for me on ubuntu 20.04 with nextcloud 22 :-(
so now what do I do?

my php version comes from `deb http://ppa.launchpad.net/ondrej/php/ubuntu focal main`

the apache config seems to refer to the php8.0 module in the mods installed, so why is this still failing?
apache -> php 8.0
alternatives -> php8.0
???

**Comment by TheColin21** [comment]
> the update-alternatives trick didn't work for me on ubuntu 20.04 with nextcloud 22 :-( so now what do I do?
> 
> my php version comes from `deb http://ppa.launchpad.net/ondrej/php/ubuntu focal main`
> 
> the apache config seems to refer to the php8.0 module in the mods installed, so why is this still failing? apache -> php 8.0 alternatives -> php8.0 ???

you may call the desired php version directly (e.g. by using php8.0 instead of php).

**Comment by pooh22** [comment]
I fixed it by removing all php packages from the server and re-installing all required php8.0- packages manually according to the documentation. Notably the apcu package drew in some php7.4 versioned packages, so I switched the config to memcached and then it worked. The nextcloud settings check suggested a few more php modules to install and now all is well again.

I'm not sure why specifically calling php8.0 (which I did try) didn't fix my problems.

The bigger question is, why did my setup fail when doing regular updates...?
I suspect it was because some php packages were installed without a specified version, which automatically drew in the latest available php packages and dependencies.

I doubt I'm the only one running into this...

**Comment by bcutter** [comment]
> I doubt I'm the only one running into this...

That's for sure. More to come and land here. E. g. because of https://github.com/nextcloud/server/issues/29287#issuecomment-1008320174.

**Comment by ruhlandm1** [comment]
> If you have both 8.0 and 8.1 on your system, it’s your job to make sure you run Nextcloud and occ (and the cron job) with 8.0. Otherwise you will get the compatibility error as expected.


Have the same problem - can someone give me advice of how to make sure that those scripts are running for sure with php 8.0? In the cronjob I´ve set it up like this: 

*/5 * * * * php8.0 -f /var/www/nextcloud/cron.php > /dev/null 2>&1

but I think this is more a workaround then a solution.

**Comment by Neodraccir** [comment]
When will 8.1 be supported? I tried to install Nextcloud yesterday and ran into this issue. Do I have to downgrade now to 8.0 (which would be a pain in the ass, since I also have to change many Nginx settings) just to upgrade it in a few weeks again?

**Comment by nickvergessen** [comment]
With the next major version Nextcloud 24.

**Comment by dvzrv** [comment]
> With the next major version Nextcloud 24.

Is there an ETA for it already? :)
For Arch Linux it is a bit of a problem to not be able to bump php a minor version unfortunately and I currently have to backport two sets of patches to somehow ensure compatibility.

**Comment by yan12125** [comment]
> Is there an ETA for it already? :)

The current plan seems the end of March - https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule#nextcloud-24

**Comment by zynexiz** [comment]
A temporary fix to get it to actually run was to comment out the version check part in lib/versioncheck.php. Doesn't seem to trow any errors or anything so far for me anyway, so it's at least up and running.

**Comment by bcutter** [comment]
> A temporary fix to get it to actually run was to comment out the version check part in lib/versioncheck.php. Doesn't seem to trow any errors or anything so far for me anyway, so it's at least up and running.

Cron?
OCC?

**Comment by zynexiz** [comment]
> > A temporary fix to get it to actually run was to comment out the version check part in lib/versioncheck.php. Doesn't seem to trow any errors or anything so far for me anyway, so it's at least up and running.
> 
> Cron? OCC?

When running the cron manually it worked, haven't tried occ yet.

**Comment by nickvergessen** [comment]
Don't do that. There can be errors along the way corrupting your data or database due to incompatible libraries and code.
If it would be only this line to change you can be sure we did it already.

**Comment by fafische** [comment]
Archlinux provides php7.4 in parallel (https://archlinux.org/news/php-80-and-php-7-legacy-packages-are-available/).
Now I am using php7.4 again because I do not want to skip php-updates for >2 month.
It is really annoying with nextcloud devellopement beeing late for every single php release..

**Comment by nickvergessen** [comment]
It would be really annoying with bugs, data loss and corruption eith every Nextcloud release because we simply include a brand new php version that has not been tested in real world with any of the used libraries and Nextcloud itself.

**Comment by CapSel** [comment]
@fafische Archlinux have nextcloud in community repository together with php. They just patched 23.0.0 release with some beta patches found here and then they happily bumped up php to 8.1.2. They call it 23.0.0-4.

It's a bit frightening that they risk stability and data just to chase newest php version.

I believe that every php application have to be migrated every time new version of php version comes out. It's just luck or ignorance that it "just works" after some checks and warnings in code are ignored.

Proper verification if everything is working well on new version of language takes time.

**Comment by zynexiz** [comment]
> @fafische Archlinux have nextcloud in community repository together with php. They just patched 23.0.0 release with some beta patches found here and then they happily bumped up php to 8.1.2. They call it 23.0.0-4.
> 
> It's a bit frightening that they risk stability and data just to chase newest php version.
> 
> I believe that every php application have to be migrated every time new version of php version comes out. It's just luck or ignorance that it "just works" after some checks and warnings in code are ignored.
> 
> Proper verification if everything is working well on new version of language takes time.

Users of Arch Linux is usually well aware (or at least should be) of what it means to run it as a server, both pros and cons. Arch Linux IS bleeding edge, it's not for everyone nor everything. For long term stability and more backward compatibility, one should use CentOS or similar distro, they are built for that.

**Comment by vvrein** [comment]
I'm on Arch, and yesterday had to switch to php7.4, since php became php8.1. (I use Nextcloud from sources). 
Usually it is normal that on Arch you have headaches sometimes, when part of your spaceship stops to work because screw hole has changed silently this morning. 😂

**Comment by zynexiz** [comment]
> I'm on Arch, and yesterday had to switch to php7.4, since php became php8.1. (I use Nextcloud from sources). Usually it is normal that on Arch you have headaches sometimes, when part of your spaceship stops to work because screw hole has changed silently this morning. joy

You could look at https://archive.archlinux.org/packages and download the previous php packages and install then, then add them to IgnorePkg in pacman.conf to prevent them to be upgraded until NextCloud supports it.

**Comment by vvrein** [comment]
>downgrade them to previous php version
Yup, I tried to, but there was some errors about missing system libraries or something like that, in which I didn't want to dig in, so just switched to php7.4.

**Comment by zynexiz** [comment]
> > downgrade them to previous php version
> 
> Yup, I tried to, but there was some errors about missing system libraries or something like that, in which I didn't want to dig in, so just switched to php7.4.

Run into that to with php-intl, missing icu version 69 library (not sure why what's the case). I installed icu69-bin from AUR and all works good now.

**Comment by txt-file** [comment]
This is not an Arch Linux Support forum. This is an issue tracking system for nextcloud.

Please move questions and help for administration of Linux distributions to the proper Linux distributions user bulletin board.

**Comment by zynexiz** [comment]
> This is not an Arch Linux Support forum. This is an issue tracking system for nextcloud.
> 
> Please move questions and help for administration of Linux distributions to the proper Linux distributions user bulletin board.

No, but a non functional NextCloud instance is an issue no matter what distro used.  Spamming a bunch of relevant posts from people trying to help out with thumbs down is pretty rude IMO.

**Comment by CapSel** [comment]
**Archlinux** users please go complain there: https://bugs.archlinux.org/task/73452 (btw I use Archlinux)

**Comment by dvzrv** [comment]
> No, but a non functional NextCloud instance is an issue no matter what distro used. Spamming a bunch of relevant posts from people trying to help out with thumbs down is pretty rude IMO.

@zynexiz This is an upstream issue tracker for the nextcloud server software. The [feature request](https://github.com/nextcloud/server/blob/a45c17dfb1da679acaa7de890bac4101f98c8892/.github/ISSUE_TEMPLATE/Feature_request.md?plain=1#L10) and [bug report](https://github.com/nextcloud/server/blob/a45c17dfb1da679acaa7de890bac4101f98c8892/.github/ISSUE_TEMPLATE/Bug_report.md?plain=1#L10) templates for tickets make its scope very clear.

I would kindly ask all Arch Linux users to use the Arch Linux [forums](https://bbs.archlinux.org/), [mailing lists](https://lists.archlinux.org/) or the `#archlinux` channel on [libera.chat](https://libera.chat), as this issue tracker is not the place to discuss issues with your installation.This ticket in particular is the place to discuss how to *ensure support* for php 8.1 in the current code base, not how to run it on your distribution (I'm packaging nextcloud for Arch Linux btw).
Not complying with the purpose of this issue tracker in general and this ticket in particular is in fact rude towards the developers of this software that made it quite clear when official support for php 8.1 will be added (see https://github.com/nextcloud/server/issues/29287#issuecomment-1018903637 and https://github.com/nextcloud/server/issues/29287#issuecomment-1019210570 for a timeline).

**Comment by zynexiz** [comment]
@dvzrv Sorry for that, didn't mean to offend anyone just tried helping out those with a broken installation.

**Comment by heyheyhello** [comment]
I'll add that a common pitfall into this mess could be the Nextcloud Updater itself. I've been running 22.2.3 and went to Admin > Overview to run the update but it says "Version 22.2.3. Nextcloud is at latest version". I know I've upgrade other servers to 23. Something must be wrong. _It's reasonable as a sysadmin (of any distro!) to think the PHP version might be preventing the upgrade_. I also checked with `php occ update:check` and `php occ upgrade` - says 22.2.3 is latest.

So. I update to PHP 8.1.

22.3.0 then tells me "This version of Nextcloud is not compatible with > PHP 8.0."

Fine, I'll manually update to 23, using `tar -xjvf`. Still the same error.

What I'm saying is Nextcloud Updater could do a lot for people to help them understand why they're being held back versions, rather than letting them debug for themselves since that can lead down the wrong path...

---

Thanks for all the amazing work you do as developers and maintainers, I know this is a complex problem and I'm grateful that it's being looked at! Just trying to suggest a way that Nextcloud could prevent people from breaking installations.

**Comment by Tyler-2** [comment]
Ouch. I've got two instances down. Wish Debian made it a touch easier to downgrade php and all of its modules.

**Comment by nursoda** [comment]
How do WordPress and similar projects manage to support new PHP versions once a new PHP version is officially released? And why is it that Nextcloud cannot? Shouldn't it have been a priority to actually have PHP 8.1 support in NC23 as it was clear that PHP 8.1 would be released while NC23 is "latest stable" and NC24 would not be available for several months?

And, yes, apart from that Linux distros should (and partly do) circumvent such incompatibilities for "slower projects", or at least [discuss solutions](https://lists.archlinux.org/pipermail/arch-dev-public/2022-January/030623.html). Given a distro's nature, the focus there is to make the distro's `nextcloud` package work. In practice, there are quite some [native NC installations](https://docs.nextcloud.com/server/latest/admin_manual/installation/source_installation.html) that cannot be helped by patching but only by either wait for the slowest to-be-supported package (using PHP) or by providing version-numbered package-sets (which is a lot of extra work).

**Comment by DjSlash** [comment]
For Debian people, with php-common version 76 both php7.4 and php8.0 can be installed on your system:
- Go to http://snapshot.debian.org/package/php-defaults/76/#php-common_2:3a:76
- Download one of the **php-common_76_all.deb** linked there
- Manually install it: `dpkg -i php-common_76_all.deb`
- Run `apt-mark hold php-common`, this will make sure the version you just installed stays the same, even if new versions are in the repos.
- (Re)Install php7.4 packages

Once NC has a release with support for 8.1, do a `apt-mark unhold php-common` and do a `apt full-upgrade`

**Comment by GivouDev** [comment]
Yeah, i dont want to complain because i mean its free software and nextcloud is freaking awesome but today a had also the mess on arch. I upgraded NextCloud via der Updater and run pacman -Syu after this....yeah and my installation of course broke...so i needed to downgrade to php-8.0.14 and then downgrade about 10 modules of php to older version which supports php 8.0.

I just hope that NextCloud has a compatible version by the next time I update my server :)

**Comment by doucettom** [comment]
Just following. I've installed PHP 8.0 and used pecl in the meantime so all extensions works but it is still a workaround. (Ubuntu Server 22.04)

**Comment by gelo2002** [comment]
I didn't see any info on this in previous comments, but the cron job is also broken due to newer PHP version.
It gives me such error when trying to run cron command:
`This version of Nextcloud is not compatible with > PHP 8.0.<br/>You are currently running 8.1.2.`

**Comment by bcutter** [comment]
Have a look at https://github.com/nextcloud/server/issues/29287#issuecomment-1014567488

Just downgrade/avoid 8.1, sit down and wait. That's what we need to do for now - and it works fine ;-)

**Comment by gelo2002** [comment]
> Have a look at [#29287 (comment)](https://github.com/nextcloud/server/issues/29287#issuecomment-1014567488)
> 
> Just downgrade/avoid 8.1, sit down and wait. That's what we need to do for now - and it works fine ;-)

Well... I'd love to downgrade, but I have another applications using php8.1
Question is, if even there will be next NC release supporting php8.1, but do to current version does not support it, there might be deadlock by upgrade, when current NC will not upgrade due to php8.1...

Wow, that was rather overly complex sentence I produced...

**Comment by ambraspace** [comment]
As many of you know, Arch Linux packagers added some patches to NC23 to force it to run with PHP 8.1. However, there's a problem with UTF-8 characters in contact names and file names.

I'm guessing it's because of this change from PHP 8.0 to PHP 8.1:

> [htmlspecialchars()](https://www.php.net/manual/en/function.htmlspecialchars.php), [htmlentities()](https://www.php.net/manual/en/function.htmlentities.php), [htmlspecialchars_decode()](https://www.php.net/manual/en/function.htmlspecialchars-decode.php), [html_entity_decode()](https://www.php.net/manual/en/function.html-entity-decode.php), and [get_html_translation_table()](https://www.php.net/manual/en/function.get-html-translation-table.php) now use ENT_QUOTES | ENT_SUBSTITUTE rather than ENT_COMPAT by default. This means that ' is escaped to &#039; while previously nothing was done. Additionally, malformed UTF-8 will be replaced by a Unicode substitution character, instead of resulting in an empty string.

For example, if you try to create txt file named 'šđčćž', you will first get an error message, but the file will be created. However, you won't be able to delete it. When you choose to delete it, it will behave as it was deleted, but when you refresh the page it will be there.

So, this happens with NC23+PHP8.1, but not with NC23+PHP8.0.

I'm writing just to draw attention to this problem so that developers can make test case (if there isn't one already).

**Comment by C0rn3j** [comment]
occ scan is also broken. 

I wish NC was tested against development PHP releases like other projects are, so that 2 months after a stable release we wouldn't be having type conversion errors breaking major functionality.

```bash
[0] # /usr/bin/runuser -u nextcloud -- /usr/bin/php /usr/share/webapps/nextcloud/occ files:scan --all
...
In Scan.php line 293:
                                                              
  Implicit conversion from float 0.15 to int loses precision                                                          
```

```php
return sprintf('%02d:%02d:%02d', ($secs / 3600), ($secs / 60 % 60), $secs % 60);
```

**Comment by smirkybg** [comment]
Changed the line to:
```php
return sprintf('%02d:%02d:%02d', ($secs / 3600), ((int)($secs / 60) % 60), $secs % 60);
````
It's good enough for the time being, until an official fix comes along.

**Comment by doucettom** [comment]
On Ubuntu 22.04 to prevent a forced upgrade to 8.1 you need to install ondrej PPA, change the apt source file to impish and install php8.0 stuff you need. I still needed to install apcu via pecl though.

**Comment by AndyScherzinger** [comment]
@come-nc just as a ping, see https://github.com/nextcloud/server/issues/29287#issuecomment-1030301870

**Comment by come-nc** [comment]
> As many of you know, Arch Linux packagers added some patches to NC23 to force it to run with PHP 8.1. However, there's a problem with UTF-8 characters in contact names and file names.
> 
> I'm guessing it's because of this change from PHP 8.0 to PHP 8.1:
> 
> > [htmlspecialchars()](https://www.php.net/manual/en/function.htmlspecialchars.php), [htmlentities()](https://www.php.net/manual/en/function.htmlentities.php), [htmlspecialchars_decode()](https://www.php.net/manual/en/function.htmlspecialchars-decode.php), [html_entity_decode()](https://www.php.net/manual/en/function.html-entity-decode.php), and [get_html_translation_table()](https://www.php.net/manual/en/function.get-html-translation-table.php) now use ENT_QUOTES | ENT_SUBSTITUTE rather than ENT_COMPAT by default. This means that ' is escaped to ' while previously nothing was done. Additionally, malformed UTF-8 will be replaced by a Unicode substitution character, instead of resulting in an empty string.
> 
> For example, if you try to create txt file named 'šđčćž', you will first get an error message, but the file will be created. However, you won't be able to delete it. When you choose to delete it, it will behave as it was deleted, but when you refresh the page it will be there.
> 
> So, this happens with NC23+PHP8.1, but not with NC23+PHP8.0.
> 
> I'm writing just to draw attention to this problem so that developers can make test case (if there isn't one already).

Is there any chance you could test this on master to see if it has the same problem? 23 does not support 8.1 and I’m not sure how much patching the ArchLinux packagers did.
Could you open a dedicated ticket for this problem with reproduction instructions and link it here?

**Comment by AndyScherzinger** [comment]
@ambraspace see https://github.com/nextcloud/server/issues/29287#issuecomment-1039969572 - could you raise a dedicated issue for this, please. Highly appreciate your efforts ❤️

**Comment by ambraspace** [comment]
@C0rn3j  @AndyScherzinger Sure. Please give me a day or two...

**Comment by ambraspace** [comment]
@come-nc @AndyScherzinger Yes, there's definitely a problem with UTF-8 with master branch and PHP 8.1.
I will open separate issue.

**Comment by ambraspace** [comment]
@come-nc @AndyScherzinger here's the link: https://github.com/nextcloud/server/issues/31212
If needed, I can upload VM image for testing.
Regards!

**Comment by jnabasny** [comment]
A temporary workaround instead of downgrading PHP system-wide is to update the cronjob to specify php8.0:

`*/5  *  *  *  * /usr/bin/php8.0 -f /var/www/html/nextcloud/cron.php`

**Comment by jnabasny** [comment]
A temporary workaround that doesn't involve downgrading the PHP version system-wide would be to specify the older version in the cronjob:

`*/5  *  *  *  * /usr/bin/php8.0 -f /var/www/html/nextcloud/cron.php`

**Comment by bcutter** [comment]
The cron is not half of the show. OCC? Updates?

Waiting is horrible I know... -.-

**Comment by gelo2002** [comment]
> A temporary workaround that doesn't involve downgrading the PHP version system-wide would be to specify the older version in the cronjob:
> 
> `*/5 * * * * /usr/bin/php8.0 -f /var/www/html/nextcloud/cron.php`

Be careful here!
Even if cron jobs will work, as @bcutter mentioned, rest of the NC is still not fully functional.
Eg. bug with non-UTF-8 characters...

**Comment by markc** [comment]
I just upgraded to Ubuntu 22.04 bumped into this issue with Nextcloud. All other PHP apps I use (Wordpress and Laravel) work fine with PHP 8.1. I'm not downgrading anything, so my not-preferred solution is to stop using Nextcloud.

**Comment by kugel-** [comment]
What's the deal with everyone?? You can co-install multiple versions of PHP on any Debian based system. And `update-alternatives` helps to manage them, perhaps by avoiding that 8.1+ is used automatically.

Nobody has to stop using nextcloud. 8.0 is easily available  without downgrade.

**Comment by markc** [comment]
I'm not prepared to install the Sury PHP packages as that then brings in alternate dependencies (openssl etc) and I risk undoing an otherwise perfectly working system for WordPress and Laravel 9 development. I could spend the time to move my Nextcloud install to another server that has not been upgrade to 22.04, but I've been through this upgrade mismatch twice before over the last 5 years, and I'm not prepared to invest any more time in this BS. My path of the least resistance is to simply stop relying on Nextcloud.

**Comment by Phillinger** [comment]
So basically you are unwilling to check the dependencies of your software before updating and unwilling to make basic adjustments after your initial error. And now it's the "BS" of Nextcloud, that forced you into that situation to make this decision? I really understand your disappointment, but come on!

**Comment by Scaenicus** [comment]
> any Debian based system

Yes. 

There is no working PHP 8.0 set (including all necessary modules) available for SuSE, only 7.4 or 8.1.
Currently I'm stuck with 7.4 and thus can't update any other application which would require 8.0+ after the upgrade. Which is okay for me, personally, but still valid for an open issue with Nextcloud.

Could we please keep this issue focused on waiting for 8.1. and not for shouting at each other. Thank you team!

**Comment by gelo2002** [comment]
You can also do a workaround by using 2 (or more) different PHP version by setting virtual hosts (I use Apache2).
I did it by hand, without any guides, but first result from google search looks exactly what's needed (though a bit older PHP and Debian versions...): [Apache + Debian](https://stackoverflow.com/questions/45033511/how-to-select-php-version-5-and-7-per-virtualhost-in-apache-2-4-on-debianl)

**Comment by darix** [comment]
> > any Debian based system
> 
> Yes.
> 
> There is no working PHP 8.0 set (including all necessary modules) available for SuSE, only 7.4 or 8.1. Currently I'm stuck with 7.4 and thus can't update any other application which would require 8.0+ after the upgrade. Which is okay for me, personally, but still valid for an open issue with Nextcloud.
> 
> Could we please keep this issue focused on waiting for 8.1. and not for shouting at each other. Thank you team!

you could find php 8.0 packages in: https://build.opensuse.org/package/show/devel:languages:php:php80/php8

**Comment by jclsn** [comment]
Nooo, I just upgraded my Debian to Bookworm. Skimmed the posts there and there seem to be workarounds for Ubuntu. I couldn't successfully install a repository with PHP8.0. Any suggestions?

**Comment by ambraspace** [comment]
> Nooo, I just upgraded my Debian to Bookworm. Skimmed the posts there and there seem to be workarounds for Ubuntu. I couldn't successfully install a repository with PHP8.0. Any suggestions?

I switched to docker installation. I will probably keep it like that even after this gets fixed.

**Comment by jclsn** [comment]
@ambraspace Will give it a shot tomorrow

**Comment by kugel-** [comment]
@jclsn install php-8.0 (that exact version) via https://deb.sury.org/

**Comment by ambraspace** [comment]
@jclsn If it helps here are docker-compose.yml and relevant part of /etc/httpd/conf/extra/httpd-vhosts-le-ssl.conf:

docker-compose.yml:
```
version: "3.7"

networks:
  default:
    ipam:
      driver: default
      config:
        - subnet: 172.20.0.0/24

services:
  nextcloud:
    networks:
      - default
    volumes:
      - "/var/lib/nextcloud/data:/var/www/html/data"
      - "/var/lib/nextcloud/config:/var/www/html/config"
      - "/var/lib/nextcloud/apps:/var/www/html/custom_apps"
    image: nextcloud
    restart: unless-stopped
    ports:
      - 8080:80
```

httpd-vhosts-le-ssl.conf:
```
<VirtualHost *:443>
    ServerName your.server.name

    SSLEngine on
    SSLCertificateFile /etc/letsencrypt/live/<domain>/fullchain.pem
    SSLCertificateKeyFile /etc/letsencrypt/live/<domain>/privkey.pem

    <IfModule mod_headers.c>
        Header always set Strict-Transport-Security "max-age=15552000; includeSubDomains; preload"
    </IfModule>

    RewriteEngine on
    RewriteRule ^/\.well-known/carddav https://%{SERVER_NAME}/remote.php/dav/ [R=301,L]
    RewriteRule ^/\.well-known/caldav https://%{SERVER_NAME}/remote.php/dav/ [R=301,L]
    RewriteRule "^$" "/" [R,L]

    ProxyRequests off

    <Location "/">
      RequestHeader set X-Forwarded-Proto 'https'

      ProxyPreserveHost On
      ProxyPass http://127.0.0.1:8080/
      ProxyPassReverse http://127.0.0.1:8080/
    </Location>
</VirtualHost>
```

Just modify it so that files and directories point to your installation.
I hope it helps.

**Comment by elyograg** [comment]
I upgraded from Ubuntu Server 20 to 22 and now I am discovering that a number of my PHP webapps are not compatible with php 8.1.  This upgrade takes PHP from 7.4 to 8.1.

**Comment by markc** [comment]
Your only option is to install the Sury deb packages and downgrade to php 8.0. The Nextcloud folks have had 6 months to get this right and still no PHP 8.1 support for one of their largest OS system targets. My personal solution was to stop using Nextcloud.

**Comment by mfb** [comment]
on the other hand at least one major opcache bug will hopefully be fixed in PHP 8.1.6 so perhaps Nextcloud won't be too far behind the curve as far as PHP 8.1 being production ready..

**Comment by come-nc** [comment]
> Nooo, I just upgraded my Debian to Bookworm. Skimmed the posts there and there seem to be workarounds for Ubuntu. I couldn't successfully install a repository with PHP8.0. Any suggestions?

Install Nextcloud 24 last RC and help us make sure that it fully support PHP 8.1?

**Comment by elyograg** [comment]
I copied the directory and the database to another server with php 7.4, and used the updater/updater.phar to upgrade in steps to version 23.  Then I copied the updated database back to my ubuntu 22 server, and tried to do "occ upgrade" after copying my config and data directory from the old version directory to version 24 rc1 that I cloned from git.  It seems to think I'm upgrading more than one major version, but the database is on version 23.  I must be missing something important.

```
elyograg@smeagol:/var/www/cloud.elyograg.org$ sudo -u www-data php occ upgrade
Nextcloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
Setting log level to debug
Exception: Updates between multiple major versions and downgrades are unsupported.
Update failed
Maintenance mode is kept active
Resetting log level
```

**Comment by elyograg** [comment]
I cannot find any instructions on how to change the release channel.  Lots of info about the dfifferent channels, but ZERO information on how to actually select a different channel.  Keep in mind that I cannot access the administration section of the app ... because it will not run on PHP 8.1.  If I try to browse to the app with version 24 rc1 (pointing at a version 23 database), I get a 504 error, with the following text:

```
The files of the app activity (activity), circles (circles), files_pdfviewer (files_pdfviewer), files_rightclick (files_rightclick), files_videoplayer (files_videoplayer), firstrunwizard (firstrunwizard), logreader (logreader), nextcloud_announcements (nextcloud_announcements), notifications (notifications), password_policy (password_policy), photos (photos), privacy (privacy), recommendations (recommendations), serverinfo (serverinfo), support (support), survey_client (survey_client), text (text), viewer (viewer) were not replaced correctly. Make sure it is a version compatible with the server.
```

My next step, which I didn't want to do because it's a lot of work, is to set up a virtualhost on the older system with php 7.4 and repoint the name there, and try to upgrade that to version 24rc1.

**Comment by elyograg** [comment]
I went ahead and did the full move of the application to the other server still running php7.4 ... and when I change the channel to beta, this is what I see at the top of the page:

```
[Nextcloud Hub II](https://nextcloud.com/) (23.0.4)

A new version is available: Nextcloud 22.2.7
```

**Comment by elyograg** [comment]
> You devs are living on another planet.

In general, I am not a developer.  I flirt with programming languages, and I think I'm actually sorta good at it, but in my career I have mostly been a system and/or network administrator.

I did finally get it all working, by completely moving the subdomain to a server still running php 7.4, upgrading it to NC 24 on that system, and then moving it all back to the primary server running Ubuntu 22.  Somewhat painful, but it worked.

**Comment by mthld** [comment]
> In general, I am not a developer.  I flirt with programming languages, and I think I'm actually sorta good at it, but in my career I have mostly been a system and/or network administrator.
> 
> 
> 
> I did finally get it all working, by completely moving the subdomain to a server still running php 7.4, upgrading it to NC 24 on that system, and then moving it all back to the primary server running Ubuntu 22.  Somewhat painful, but it worked.

I'm glad you got it working! It may haven't been clear but I had something against @come-nc answer, not yours. 😊

**Comment by Congruentie** [comment]
I just did a clean install of the latest production Ubuntu 22.04 LTS, Apache2, MySQL, PHP8.1.4 and all the mods I had to install for NextCloud as indicated by the web installer. That didn´t work, for otherwise I would not have found this forum ;). I am glad I did though. So next I installed the latest RC NextCloud 23.0.4 March 21 2022, to find the same result.
I am pretty new to Ubuntu, but a fast learner. GitHub is completely new to me, so still figuring out where to find Version 24 ;), cause I couldn´t find it on the website. Willing to spend time figuring things out though.

**Comment by come-nc** [comment]
You should be able to switch release channel in the configuration page, or through the config file.
Documentation about this is here: https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#nextcloud-verifications
More details about release channels: https://nextcloud.com/release-channels/

> I did finally get it all working, by completely moving the subdomain to a server still running php 7.4, upgrading it to NC 24 on that system, and then moving it all back to the primary server running Ubuntu 22. Somewhat painful, but it worked.

Glad to see you found a way out of the locked down situation.

**Comment by jancborchardt** [comment]
Hi @mthld, please keep a friendly and constructive tone.
Nextcloud is an open source project & community, so everyone is participating – it's not just a few people developing for an audience. Thank you! :)

**Comment by J4gQBqqR** [comment]
> I did finally get it all working, by completely moving the subdomain to a server still running php 7.4, upgrading it to NC 24 on that system, and then moving it all back to the primary server running Ubuntu 22. Somewhat painful, but it worked.

I went through a more painful experience by upgrading my Ubuntu 20 to Ubuntu 22. Nextcloud suddenly disconnects. Fearing that I am hacked and losing data, unplugged my Ethernet cable...

Suspect that the issue is due to php upgraded to 8.1 in Ubuntu, first tried php8.0 as it is marked as _recommended_ in the documentation.
It does not work. So many required modules cannot be found for php8.0 ... Does anyone know why? If 8.0 has so little modules available, why is it _recommended_ in the official documentation?

Then uninstall the whole php8 and went back to 7.4.
Note that I did all these via USB drive without ethernet pluggin in. Very painful...

@jancborchardt I understand your point about not developing for a single audience. But please note that a potentially significant percentage of the audiences are running Ubuntu and will upgrade to Ubuntu 22. 

Potentially a large number of audiences will not know Nextcloud will fail until they landed in Ubuntu 22 and starting to research for the point of failure and landed here.

Please, please, discuss internally within Nextcloud developers about the priority of Nextcloud support for php8.1, not in long extended future release of 24, but in the current release of 23. Hope that the community's feedback can be heard.

**Comment by jens-t** [comment]
Nextcloud server 24.0.0 stable is released. :tada:  No problems with php8.1.

Did i miss anything or can this issue be closed now?

**Comment by pooh22** [comment]
Hi Jens, though I'm not the one who posted this bugreport, I think this issue is/has become more about quality and support than whether _a version_ of nextcloud supports php8.1.
The original poster mentioned a month in advance about the upcoming php 8.1 release and how this would affect all nextcloud admins running NC22. Not providing a timely fix for all supported NC versions for php8.1 has been a dissapointment for all posters in this issue, I'm sure. Since I'm not a developer, I don't know exactly why this was such a big problem, but as sysadmin, I want to be sure all software I run is safe to run on the latest stable OS releases.

I'm very grateful for all people working on NC to provide it and all the features and plugins that I love to use. And I know it's open source and as such can be a bit of a rough ride sometimes. Still the way this particular issue was handled makes me want to look elsewhere for the same functionality (haven't found it though).

For my part, I would close this when the php 8.1 support is backported to at least NC23 and not before that. Also I would love some re-assurance that this kind of FU will not be happening again anytime soon.

**Comment by jclsn** [comment]
Yeah, next problem: Nextcloud can't update from 22 to 24. Since 23 doesn't work with PHP8.1, I have to completely reinstall...

**Comment by seuffert** [comment]
> Yeah, next problem: Nextcloud can't update from 22 to 24. Since 23 doesn't work with PHP8.1, I have to start from scratch...

I had the same problem. Download nc24 and update manually on command line.

**Comment by jclsn** [comment]
@seuffert : How? That is what I tried. 

```
 ⚡ root@miniserver  /var/www/html/nextcloud  sudo -u www-data php occ upgrade
Nextcloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
Setting log level to debug
Exception: Updates between multiple major versions and downgrades are unsupported.
Update failed
Maintenance mode is kept active
Resetting log level
```

**Comment by unilock** [comment]
@jclsn @seuffert Upgrading manually via the command line still requires incremental upgrades between major versions IIRC. The only(?) solution would be to set up a (virtual?) system with PHP 7.4 or 8.0, backup and move your Nextcloud configuration to it, use that setup to migrate to 22 -> 23 -> 24, then move your Nextcloud config back to the original system.

**Comment by jclsn** [comment]
@unilock Not on a Saturday. 

I will use the AIO docker version now. It seems to work this time.

**Comment by elyograg** [comment]
> The only(?) solution would be to set up a (virtual?) system with PHP 7.4 or 8.0, backup and move your Nextcloud configuration to it, use that setup to migrate to 22 -> 23 -> 24, then move your Nextcloud config back to the original system.

This is what I did.  I changed where my front end haproxy was directing that domain, copied the database and installation files to another machine still running PHP 7.4, upgraded it to 24 there, made sure it was working, and then moved it back to the server I had upgraded to Ubuntu 22.

Upgrading the OS broke most of my PHP-based web applications.  Seems very few projects were prepared for changes in PHP 8.  I've been commenting on bug reports all over the place trying to get things fixed.

One website, written for uwsgi, I couldn't make it run at all.  I found some PHP replacements, but I couldn't get any of those working with PHP 8.1, so I picked one of them, installed it on that older system, and now it works.

The system still running the older PHP version is a NanoPi.  https://raspi2.elyograg.org/  I have upgraded all my servers except that one to Ubuntu 22.  Who knows how long before the ubuntu-derived OS I'm running on the NanoPi will have an upgrade available.  I may not upgrade it for quite a while, because it is a last refuge where I can run apps that I simply cannot make work with PHP 8.

**Comment by C0rn3j** [comment]
@ChristophWurst  Did anything actually change to prevent this situation in the future, or will this exact same situation play out in the future with 8.2?

Is the Nextcloud team going to track PHP development/beta releases and make sure current major Nextcloud releases get support in patch versions?

Please note the huge amount of likes and the amount of participants in this thread.

**Comment by AndyScherzinger** [comment]
We are aware that there are many likes on this topic but I am afraid you will have to live with the risk that this might happen again in the future for certain scenarios, the ones we also faces with 8.1. There have been 3rd party libraries we use, i.e. sabre that haven't been compatible with 8.1 in the recent past. This basically makes it impossible then to support 8.1 right from the initial release.

I understand that anybody who can would love to simply always run everything at latest-and-greatest, I am the first in line here. However we do state in the requirements with which versions of the underlying system/platform we are compatible.

The question with the patch version is hard to answer, we did our best to get it compatible with 24.0.0 and if there is something still not right we would fix in a patch/minor release of 24 since that is the latest major release however we do not backport new version support to older Nextcloud version to keep changes within a major release rather small as in mostly fixes. This implies that new PHP version support is something I do not consider a bug and thus only targets the latest release.

So I keep my fingers crossed looking at 8.2 of PHP that we won't run into the same scenario and all the used libraries will support it right from the start but there is no guarantee. 

I hope this sheds some light on the how and why things tunrned out the way they did.

**Comment by jamiew0w** [comment]
hi guys and girls. is there any plans to help people stuck on nc22 upgrade to nc24? (IE an easier upgrade path)

**Comment by nanaya** [comment]
I haven't tried this but if you can get older php binary somehow in the system (phpenv, etc), maybe try running `/path/to/bin/php7.4 updater/updater.phar` from cli to update to version 24? (or whatever the php version and path)

**Comment by rdlab-upc** [comment]
@jclsn @elyograg @jamiew0w @C0rn3j et all

Maybe this workaround will help for Ubuntu 22.04 and Nextcloud 22/23/24, at least our tests worked fine and clean. Ondrej (now) provides php 8.0 packages for Jammy 22.04 (https://launchpad.net/~ondrej/+archive/ubuntu/php)

**HowTo add php8.0 under Ubuntu 22.04**

1. Add Ondrej repository
sudo add-apt-repository ppa:ondrej/php

2. Update sources

sudo apt update

3. Install php8.0 with desired modules

apt-get install php8.0 php8.0-cli php8.0-fpm .....

4. (OPTIONAL) make php8 default in your system (for cron/cli executions)

sudo update-alternatives --config php

P.S: Probably you need to change/restart the fastcgi_pass unix:/run/php/php-fpm.sock;  in your apache/nginx configuration to point out to /var/run/php/php8.0-fpm.sock

P.S2: You can also delete all php-8.1 system packages if no more system dependencies exist, but IMHO is better having both versions available in the system.

@AndyScherzinger @pooh22 First of all a big thank you and real acknowledge to the NC team, great job!.  It is plain to see the big effort you made this past months.

Said so, IMHO, this issue could be addressed more politely from the developer team. (some ;) Claims are nicely motivated and explained and replies like "So I keep my fingers crossed looking at 8.2 of PHP that we won't run into the same scenario and all the used libraries will support it right from the start but there is no guarantee." do not provide trustworthiness in this project.

I know and understand that this is a community initiative lead by NC staff, so the community (we would like to believe we belong to) must actively contribute and not only complain about "my own needs". This is why we share this "howto" in order to help address this problem.

Sorry for the looong post ;)

Thank you!

**Comment by notandxor** [comment]
Got bitten by this by a fedora 36 upgrade. Downgardig to php 8.0,  this saved my day:
https://rpms.remirepo.net/wizard/

**Comment by xxblx** [comment]
Thanks @notandxor! It's also possible to install php 8.0 as a separate version and use it for upgrading only, i.e., without downgrading php system-wide. 

Workaround for Fedora 36 
```
dnf install https://rpms.remirepo.net/fedora/remi-release-36.rpm
dnf config-manager --set-enabled remi
dnf install php80 php80-php-pgsql php80-php-pecl-apcu php80-php-intl php80-php-intl php80-php-gd php80-php-xml php80-php-pdo php80-php-process php80-php-pecl-zip php80-php-mbstring
```
Do manual upgrade with php 8.0. I applied `--define` as I didn't configure php80 because I was not going to keep it after the upgrade.
```
php80 --define apc.enable_cli=1 occ upgrade
```
I upgraded 22.2.5 => 22.2.8 => 23.0.5 with php80 from the remi repository and after that upgraded to 24.0.1 using default php from the fedora repository. So far 24.0.1 has been working fine with php 8.1.

**Comment by BlackLight** [comment]
I've kept my `php*` packages on hold on Arch Linux for way too long (while waiting for NC 24 to be rolled out, hopefully with PHP 8.1 support). And this is starting to affect other `occ` commands that can no longer find the versions of the libraries that they expect:

```
PHP Warning:  PHP Startup: Unable to load dynamic library 'intl' (tried: /usr/lib/php/modules/intl (/usr/lib/php/modules/intl: cannot open shared object file: No such file or directory), /usr/lib/php/modules/intl.so (libicudata.so.70: cannot open shared object file: No such file or directory)) in Unknown on line 0
```

Note that Arch Linux doesn't support partial upgrades, so this was expected to happen at some point. I just hoped that I didn't have to hold on the PHP system packages for so long for the system itself to start falling apart around it.

How am I supposed to tackle this when the upgrade to NC 24 comes to my instance, since `occ` commands are most likely to fail because some of the dependencies that they rely on have been upgraded?

**Comment by heyheyhello** [comment]
@BlackLight you should not have needed to hold back PHP versions, as Arch patched their Nextcloud 23 to support PHP 8.1 months ago beginning in January  https://github.com/archlinux/svntogit-community/commit/c78893ca625b81a48d3201ac31dff70b2477133f

**Comment by zynexiz** [comment]
@BlackLight Had my PHP on hold to for a while, but upgraded to v24.0.0 a week or two ago, released the hold and upgraded everything. It all works very good again now, and think actually NC got a bit faster (feels snappier anyway). Just updated to 24.0.1 yesterday, still running good. I don't use NC package from repo thou because of my structure on my web server and stuff.

**Comment by BlackLight** [comment]
> @BlackLight you should not have needed to hold back PHP versions, as Arch patched their Nextcloud 23 to support PHP 8.1 months ago beginning in January [archlinux/svntogit-community@c78893c](https://github.com/archlinux/svntogit-community/commit/c78893ca625b81a48d3201ac31dff70b2477133f)

I don't have the NC package installed via pacman. I installed it directly in my Apache directory because of quirks/customizations with my server. And the upgrade to NC 24 hasn't yet come in my mailbox. And when it comes, it's likely to break because of the mismatch between expected/available dependencies of PHP. What shall I do in this case? A few options:

- Install the `php80*` AUR versions for all the PHP packages?
- Upgrade the system `php` packages and cross my fingers?
- Do a fresh installation?

**Comment by zynexiz** [comment]
@BlackLight You should be able to upgrade true the UI (or shell), v24.0.0 has been there for a while now (and also v24.0.1). I did my upgrade a couple of weeks ago. Can confirm that it works with latest PHP in arch repo. I upgraded NC first, then released the packages and did pacman -Syyu on everything.

**Comment by nariox** [comment]
On Arch, php7 is still part of the official repos. I keep both of them
installed just in case. Not the most minimalistic, but they tend to upgrade
php really quickly, so it's a good idea to have it installed just in case.

**Comment by BlackLight** [comment]
@AndyScherzinger thanks for the instructions. I have eventually managed to upgrade manually (downloaded the latest zip to the Apache dir and `occ upgrade`). Then after upgrading to NC 24 I upgraded PHP to 8.1 and everything went fine. I was afraid that some of the missing system dependencies would break the `occ upgrade` process, but luckily the only PHP module that was broken (because it relied on the `libcuuc*` libraries that were upgraded) was `intl`, which apparently is not mandatory for the upgrade to run.

**Comment by AndyScherzinger** [comment]
My pleasure @BlackLight and happy to hear you arrived on 24 safely 👍

**Comment by VVD** [comment]
Several errors:
```
strtolower(): Passing null to parameter #1 ($string) of type string is deprecated at /nextcloud/3rdparty/swiftmailer/swiftmailer/lib/classes/Swift/Transport/EsmtpTransport.php#144
```
3rdparty/swiftmailer/swiftmailer/lib/classes/Swift/Transport/EsmtpTransport.php#144:
```
-$encryption = strtolower($encryption);
+$encryption = strtolower($encryption ?? '');
```
```
str_replace(): Passing null to parameter #3 ($subject) of type array|string is deprecated at /nextcloud/apps/mail/vendor/bytestream/horde-mime/lib/Horde/Mime/Headers/Element.php#145
```
apps/mail/vendor/bytestream/horde-mime/lib/Horde/Mime/Headers/Element.php#145:
```
-return str_replace("\0", '', $data);
+return str_replace("\0", '', $data ?? '');
```

**Comment by AndyScherzinger** [comment]
@VVD please raise this at https://github.com/nextcloud/mail/issues including the version infos server/mail-app

 - also cc @ChristophWurst since it is related to mail

**Comment by ChristophWurst** [comment]
> 3rdparty/swiftmailer

https://github.com/nextcloud/3rdparty / https://packagist.org/packages/swiftmailer/swiftmailer

> bytestream/horde-mime

https://github.com/bytestream/Mime

**Comment by VVD** [comment]
Thanks for information.

> > 3rdparty/swiftmailer
> 
> https://github.com/nextcloud/3rdparty / https://packagist.org/packages/swiftmailer/swiftmailer

```
This package is abandoned and no longer maintained.
The author suggests using the [symfony/mailer](https://packagist.org/packages/symfony/mailer) package instead.
```
Found commit with PHP8.1 support: https://github.com/swiftmailer/swiftmailer/commit/ee8a1d9684ba571e4c0f9b857b1a81a9c419186b
It have "my" patch too.

> > bytestream/horde-mime
> 
> https://github.com/bytestream/Mime, addressed in [bytestream/Mime#3](https://github.com/bytestream/Mime/pull/3)

Already fixed, but when we can see it in our installation? What are the plans?

**Comment by AndyScherzinger** [comment]
@VVD can you please raise a new issue for this one, thanks, your effort is highly appreciated, cc @PVince81

**Comment by VVD** [comment]
@AndyScherzinger, https://github.com/nextcloud/server/issues/33295

**Comment by creopard** [comment]
> Several errors:
> 
> ```
> strtolower(): Passing null to parameter #1 ($string) of type string is deprecated at /nextcloud/3rdparty/swiftmailer/swiftmailer/lib/classes/Swift/Transport/EsmtpTransport.php#144
> ```
> 
> 3rdparty/swiftmailer/swiftmailer/lib/classes/Swift/Transport/EsmtpTransport.php#144:
> 
> ```
> -$encryption = strtolower($encryption);
> +$encryption = strtolower($encryption ?? '');
> ```
> 
@ChristophWurst (cc @AndyScherzinger )
The above issue with swiftmailer found by @VVD is still open in Nextcloud 25.0.1 and pops up regularely in the NC error log when sending an email.

What about the existing fixes https://github.com/swiftmailer/swiftmailer/commit/ee8a1d9684ba571e4c0f9b857b1a81a9c419186b that were not yet integrated into https://github.com/nextcloud/3rdparty/tree/master/swiftmailer/swiftmailer ?

**Comment by ChristophWurst** [comment]
Could you open that as a separate ticket please @creopard? Then one of the server engineers can have a look.
